include/swift/SIL/Dominance.h

```diff
@@ -184,6 +185,62 @@ class PostDominanceInfo : public PostDominatorTreeBase {
   using super::properlyDominates;
 };
 
+/// Invoke the given callback for all the reachable blocks
+/// in a function.  It will be called in a depth-first,
+/// dominance-consistent order.
+///
+/// Furthermore, prior to running each block, a tracking scope will
+/// be entered for each of the trackers passed in, as if by:
+///
+///   typename ScopedTrackingTraits<Tracker>::scope_type scope(tracker);
+///
+/// This allows state to be saved and restored for each of the trackers,
+/// such that each tracker will only represent state that was computed
+/// in a dominating block.
+template <class... Trackers, class Fn>
+void runInDominanceOrderWithScopes(DominanceInfo *dominance, Fn &&fn,
+                                   Trackers &...trackers) {
+  using TrackingStackNode = TrackingScopes<Trackers...>;
+  llvm::SmallVector<std::unique_ptr<TrackingStackNode>, 8> trackingStack;
+
+  // The stack of work to do.  A null item means to pop the top
+  // entry off the tracking stack.
+  llvm::SmallVector<DominanceInfoNode *, 16> workStack;
+  workStack.push_back(dominance->getRootNode());
+
+  while (!workStack.empty()) {
+    auto node = workStack.pop_back_val();
+
+    // If the node is null, pop the top entry off the tracking stack.
+    if (node == nullptr) {
+      (void) trackingStack.pop_back_val();
+      continue;
+    }
+
+    auto bb = node->getBlock();
+
+    // If the node has no children, build the stack node in local
+    // storage to avoid having to heap-allocate it.
+    if (node->isLeaf()) {
+      TrackingStackNode stackNode(trackers...);
+
+      fn(bb);
+
+    // Otherwise, we have to use the more general approach.
+    } else {
+      // Push a tracking stack node.
+      trackingStack.emplace_back(new TrackingStackNode(trackers...));
```

Comment:
You only need to push a node for blocks which actually contain any hop_to_executor instructions.
Maybe you can extend this API so that pushing a tracking stack node can be done conditionally?

Reply:
I think that, if we can optimize this to not require heap-allocation, or at least to require fewer allocations, that's better than trying to pursue micro-optimizations based on whether trackers actually have values inserted into them.

Reply:
fine with me